controllers/src/main/java/io/syndesis/controllers/integration/online/ActivateHandler.java

```diff
@@ -318,4 +350,36 @@ private static boolean hasComponentProperties(Map<String, String> properties, Wi
 
         return false;
     }
+
+    // ===============================================================================
+    // Some helper method to conditional execute certain steps
+    @FunctionalInterface
+    public interface IoCheckedFunction<T> {
+         void apply(T t) throws IOException;
+    }
+
+    private class BuildStepPerformer {
+        private final List<String> stepsPerformed;
+        private Integration integration;
+
+        BuildStepPerformer(Integration integration) {
+            this.integration = integration;
+            this.stepsPerformed = integration.getStepsDone().orElseGet(ArrayList::new);
```

Comment:
This list will be unmodifiable if coming from `integration`, so stepsPerformed.add below might fail.

Reply:
good catch, hadn't yet a situation where one of the steps failed. I hope to get rid of keeping such kind of state on the integration in the near future, though.